http/peer-id-auth.md

```diff
@@ -0,0 +1,101 @@
+# Peer ID Authentication over HTTP
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+| --------------- | ------------- | ------ | --------------- |
+| 1A              | Working Draft | Active | r0, 2023-01-23  |
+
+Authors: [@MarcoPolo]
+
+Interest Group: Same as [HTTP](README.md)
+
+## Introduction
+
+This spec defines one way of authenticating Peer IDs over HTTP using a
+challenge-response scheme.
+
+## Mutual Client and Server Peer ID Authentication
+
+1. The server initiates the authentication by responding to a request that must
+   be authenticated with the response header `WWW-Authenticate: Libp2p-Challenge
+   challenge="<base64-encoded-challenge>, Libp2p-Challenge-Server-Only"`. The
+   challenge MUST be randomly generated from server for sole purpose of
+   authenticating the client. The server SHOULD store the challenge temporarily
+   until the authentication is done. The challenge SHOULD be at least 32 bytes.
+1. The client sends a request and sets the `Authorization`
+   [header](https://www.rfc-editor.org/rfc/rfc9110.html#section-11.6.2) header
+   to the following:
+   ```
+   Libp2p-Challenge peer-id="<encoded-peer-id-bytes>",client-challenge="<base64-encoded-client-challenge>",sig="<base64-signature-bytes>"
+   ```
+   * The peer-id is encoded per the [peer-ids spec](../peer-ids/peer-ids.md).
+   * The signature is over the concatenated result of:
+   ```
+     <varint-length> + "origin=" + server-name + 
+     <varint-length> + "client-challenge=" + base64-encoded-client-chosen-client-challenge + 
+     <varint-length> + "challenge=" + base64-encoded-challenge
+   ```
+   * The client chosen client-challenge MUST be randomly generated.
+   * The client chosen client-challenge SHOULD be at least 32 bytes.
+   * The client MUST use the same server-name as what is used for the TLS
```

Comment:
Can this now be used on subsequent requests?

Reply:
Maybe? I'm not sure. I could be better to have a separate scheme for that. Like the previous libp2p token.

Reply:
Let's use the bearer token [RFC 6750](https://datatracker.ietf.org/doc/html/rfc6750) for this. It's well defined and is pretty much what we want. Essentially, it's uses the request header of "Authorization: Bearer <token>".

I'll update the spec to include information on how the server can return a bearer token.

Reply:
I've updated the spec to include a blurb about how the server may return a bearer token.